ncm-opennebula/src/main/perl/OpenNebula/Ceph.pm

```diff
@@ -0,0 +1,129 @@
+
+C<NCM::Component::OpenNebula::Ceph> adds C<Ceph> backend support to
+L<NCM::Component::OpenNebula::Host>.
+
+=head2 Public methods
+
+=over
+
+=item enable_ceph_node
+
+Configures C<Ceph> client and
+set the C<Ceph> key in each host.
+
+=cut
+
+sub enable_ceph_node
+{
+    my ($self, $type, $host, $datastores) = @_;
+
+    foreach my $ceph (sort keys %$datastores) {
+        my ($secret, $uuid);
+        if ($datastores->{$ceph}->{tm_mad} eq 'ceph') {
+            if ($datastores->{$ceph}->{ceph_user_key}) {
+                $self->verbose("Found Ceph user key in $ceph datastore");
+                $secret = $datastores->{$ceph}->{ceph_user_key};
+            } else {
+                $self->error("Ceph user key not found within $ceph datastore.");
+                return;
+            }
+            $uuid = $self->set_ceph_secret($type, $host, $datastores->{$ceph}->{ceph_secret});
+            if (!$uuid) {
+                $self->error("Not able to set libvirt uuid for $ceph datastore");
+                return;
+            }
+            if (!$self->set_ceph_keys($host, $uuid, $secret)) {
+                $self->error("Not able to set ceph libvirt key for $ceph datastore");
+                return;
+            }
+        }
+    }
+    return 1;
+}
+
+=item set_ceph_secret
+
+Sets the C<Ceph> secret to be used by C<libvirt>.
+
+=cut
+
+sub set_ceph_secret
+{
+    my ($self, $type, $host, $ceph_secret) = @_;
+    my $uuid;
+    # Add ceph keys as root
+    my $cmd = ['secret-define', '--file', $CEPHSECRETFILE];
+    my $output = $self->run_virsh_as_oneadmin_with_ssh($cmd, $host);
+    if ($output and $output =~ m/^[Ss]ecret\s+(.*?)\s+created$/m) {
+        $uuid = $1;
+        if ($uuid eq $ceph_secret) {
+            $self->verbose("Found Ceph uuid $uuid to be used by $type host $host in $CEPHSECRETFILE.");
+        } else {
+            $self->error("UUIDs set from datastore and CEPHSECRETFILE $CEPHSECRETFILE do not match.");
+            return;
+        }
+    } else {
+        $self->error("Required Ceph UUID not found for $type host $host.");
+        return;
+    }
+    return $uuid;
+}
+
+=item set_ceph_keys
+
+Sets the C<Ceph> keys to be used by C<libvirt>.
+
+=cut
+
+sub set_ceph_keys
+{
+    my ($self, $host, $uuid, $secret) = @_;
+
+    my $cmd = ['secret-set-value', '--secret', $uuid, '--base64', $secret];
+    my $output = $self->run_virsh_as_oneadmin_with_ssh($cmd, $host, 1);
+    if ($output =~ m/^[sS]ecret\s+value\s+set$/m) {
+        $self->info("New Ceph key include into libvirt $host uuid $uuid list: ", $output);
+    } else {
+        $self->error("Error running virsh secret-set-value command for $host uuid $uuid: ", $output);
+        return;
+    }
+    return $output;
+}
+
+=item detect_ceph_datastores
+
+Detects any OpenNebula C<Ceph> datastore setup.
+
+=cut
+
+sub detect_ceph_datastores
+{
+    my ($self, $one) = @_;
+    my @datastores = $one->get_datastores();
+    
+    foreach my $datastore (@datastores) {
```

Comment:
this can all be replaced by
```perl
my @cephs = grep {$_->{data}->{TM_MAD}->[0] eq "ceph"} @{$one->get_datastores()};
if (@cephs) {
    $self->verbose("Detected Ceph datastore(s): ", join(', ', @cephs));
} else {
    $self->verbose("No Ceph datastores were found");
}
return @cephs ? 1 : undef;
```